Bluetooth: Host: Update BR L2CAP channel state

lylezhu2012 opened the pull request:
In the function bt_l2cap_chan_set_state, BR channel is treated as LE Chan.
The BR L2CAP channel state is updated incorrectly.

Add a function bt_l2cap_br_chan_set_state_debug for BR L2CAP channel. Call the function when the connection is BR in bt_l2cap_chan_set_state.

carlescufi reviewed (dismissed):
Do we really need the second patch reformatting the code?
@jhedberg @jori-nordic

jori-nordic commented:
@carlescufi wrong PR? I don't see formatting changes.

jori-nordic commented on subsys/bluetooth/host/l2cap_br_interface.h:
```diff
@@ -37,3 +37,14 @@ void l2cap_br_encrypt_change(struct bt_conn *conn, uint8_t hci_status);
 /* Handle received data */
 void bt_l2cap_br_recv(struct bt_conn *conn, struct net_buf *buf);
+
+#if defined(CONFIG_BT_L2CAP_LOG_LEVEL_DBG)
```
I don't think you need the guard here.

lylezhu2012 replied:
If CONFIG_BT_L2CAP_LOG_LEVEL_DBG is defined, the symbol bt_l2cap_br_chan_set_state is macro. And it routes to  bt_l2cap_br_chan_set_state_debug. Or, the symbol bt_l2cap_br_chan_set_state is a function. 

So, the preprocessor condition is used to avoid duplicate definitions the symbol bt_l2cap_br_chan_set_state.

carlescufi commented:
> @carlescufi wrong PR? I don't see formatting changes.

No, right PR but @lylezhu2012 removed the commit :)

lylezhu2012 commented:
> > @carlescufi wrong PR? I don't see formatting changes.
> 
> No, right PR but @lylezhu2012 removed the commit :)

Yes. I removed  the clang-format commit.

I found the conflict in my clang-format commit. So I plan to recommit it.

And I found it is huge change. It is hard to maintain and blame. 

> @carlescufi wrong PR? I don't see formatting changes.

And refer to comments from @carlescufi, I removed it from this PR.

Do I need to format these files again?

jori-nordic commented:
> Do I need to format these files again?

I wouldn't, if not really necessary. As you said it makes it harder to use git-blame.

jhedberg commented on subsys/bluetooth/host/l2cap_br.c:
```diff
@@ -704,6 +704,62 @@ static int l2cap_br_conn_req_reply(struct bt_l2cap_chan *chan, uint16_t result)
+		/* regardless of old state always allows this state */
+		break;
+	case BT_L2CAP_CONNECTING:
+		if (br_chan->state != BT_L2CAP_DISCONNECTED) {
+			LOG_WRN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_L2CAP_CONFIG:
+		if (br_chan->state != BT_L2CAP_CONNECTING) {
+			LOG_WRN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_L2CAP_CONNECTED:
+		if (br_chan->state != BT_L2CAP_CONFIG &&
+		    br_chan->state != BT_L2CAP_CONNECTING) {
+			LOG_WRN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_L2CAP_DISCONNECTING:
+		if (br_chan->state != BT_L2CAP_CONFIG &&
+		    br_chan->state != BT_L2CAP_CONNECTED) {
+			LOG_WRN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	default:
+		LOG_ERR("%s()%d: unknown (%u) state was set", func, line, state);
```
I think it's discouraged to manually insert function/line information into logging strings, since the same information can be enabled through the logging subsystem's build configuration. The line numbers aren't even that useful here - what you really want is the before and after states, i.e. I'd suggest to use a "state2str()" conversion helper to log the invalid transitions as human readable strings instead of numbers.

lylezhu2012 replied:
Removed parameters line and func.

jhedberg replied:
@lylezhu2012 ah, I see now, sorry. I thought the code was logging the line/func of the LOG_*() macro, but this code exists to instead log the place where `set_state()` was called from. You can disregard this review comment then.

jhedberg reviewed (changes requested):
See my inline comment

lylezhu2012 committed:
Bluetooth: Host: Add bt_l2cap_br_disconnected

Handle BR/EDR L2CAP disconnection notification
separately.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Host: Add bt_l2cap_br_chan_del

Add bt_l2cap_br_chan_del for BR/EDR channel.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jori-nordic commented on subsys/bluetooth/host/l2cap_br_interface.h:
```diff
@@ -37,3 +40,17 @@ void l2cap_br_encrypt_change(struct bt_conn *conn, uint8_t hci_status);
 
 /* Handle received data */
 void bt_l2cap_br_recv(struct bt_conn *conn, struct net_buf *buf);
+
```
could you move this whole chunk back to `l2cap_br.h`?
I'd like to keep `l2cap_br_interface.h` only for functions that are called from `l2cap.c`.

lylezhu2012 replied:
I did not find file `l2cap_br.h`. Do you mean it is file `l2cap_br_internal.h`? If yes, I think we can move line 44 ~ line 56 to file `l2cap_br_internal.h`.

lylezhu2012 committed:
Bluetooth: Host: Update BR L2CAP channel state

In the function bt_l2cap_chan_set_state, BR
channel is treated as LE Chan.
The BR L2CAP channel state is updated
incorrectly.

Add a function bt_l2cap_br_chan_set_state
for BR L2CAP channel. For BR/EDR channel,
call bt_l2cap_br_chan_set_state directly.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented on subsys/bluetooth/host/l2cap_br.c:
```diff
@@ -715,6 +715,37 @@ static int l2cap_br_conn_req_reply(struct bt_l2cap_chan *chan, uint16_t result)
+	chan->conn = NULL;
+
+destroy:
+#if defined(CONFIG_BT_L2CAP_DYNAMIC_CHANNEL)
```
Isn't dynamic channels in practice implied if you have BR/EDR enabled? Even SDP uses dynamic channels (PSM 1). If this isn't already enforced in Kconfig it probably should be.

jori-nordic replied:
yeah I think we definitely need to touch-up the dynamic-channel compilation scope. 
There are redundant `IS_ENABLED()` and `#ifdef`s. We could probably move everything related to dynamic channels to their own conditionally-compiled files.

Regarding the enforcement, I think @lylezhu2012 can make another (trivial) PR if that's ok.

jhedberg replied:
> Regarding the enforcement, I think @lylezhu2012 can make another (trivial) PR if that's ok.

Sure. That's fine with me.
